### PR TITLE
Tests for Keycloak Deployment

### LINKED ISSUE #9732: Tests for Keycloak Deployment
This is a follow-up for #9171

### PR DESCRIPTION
**Pull request by vmuzikar**
Fixes #9732

### CODE REVIEW

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    private void waitForKcToFail(Keycloak kc) {
+        Awaitility.await()
+                .ignoreExceptions()
+                .pollDelay(Duration.ofSeconds(60)) // just to be sure the cluster had enough time to deploy KC
+                .untilAsserted(() -> {
+                    var currentKc = k8sclient.resources(Keycloak.class).withName(kc.getMetadata().getName()).get();
+                    CRAssert.assertKeycloakStatusCondition(currentKc, KeycloakStatusCondition.READY, false);
+                    // we currently don't detect ImagePullBackOff, pod, config, ... errors in the operator
+                    // TODO https://github.com/keycloak/keycloak/issues/10099
+                    // CRAssert.assertKeycloakStatusCondition(currentKc, KeycloakStatusCondition.HAS_ERRORS, true);
```
Comment (andreaTP) [inline operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:128]: Please remove commented code, we already have an issue tracking the problem.

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    @Test
+    public void testImageChange() {
+        try {
+            var kc = getDefaultKeycloakDeployment(k8sclient);
+            kc.getSpec().setImage("quay.io/keycloak/non-existing-keycloak");
+
+            deployKeycloak(k8sclient, kc, false);
+            waitForKcToFail(kc);
```
Comment (andreaTP) [inline operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:59]: This check is not quite enough to tell that it's failing for `ImagePullBackoff`, either:
 - remove the test at all and tackle it in https://github.com/keycloak/keycloak/issues/10099
 - check for the specific condition

File: operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java [hunk 3]
```diff
@@ -122,6 +137,41 @@ private static void calculateNamespace() {
...
+    k8sclient.load(KeycloakDeploymentE2EIT.class.getResourceAsStream("/example-postgres.yaml")).inNamespace(namespace).createOrReplace();
+
+    // Check DB has deployed and ready
+    Log.info("Checking Postgres is running");
+    Awaitility.await()
+            .untilAsserted(() -> assertThat(k8sclient.apps().statefulSets().inNamespace(namespace).withName("postgresql-db").get().getStatus().getReadyReplicas()).isEqualTo(1));
+  }
+
+  // TODO improve this (preferably move to JOSDK)
+  protected void savePodLogs() {
```
Comment (andreaTP) [inline operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java:152]: this forces us to enclose all the tests in big `try ... catch` blocks, it will also increase the technical debt(already pretty high) of the tests.
Since this is not related to the scope of this PR, can you please exclude from the current PR and open a relevant ticket?

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                        assertThat(d.getMetadata().getLabels().entrySet().containsAll(labels.entrySet())).isTrue(); // additional labels should not be overwritten
+                        assertThat(d.getSpec()).isEqualTo(origSpecs); // specs should be reconciled back to original values
+                    });
+        } catch (Exception e) {
+            savePodLogs();
+            throw e;
+        }
+    }
+
+    // TODO Test HasErrors condition; ReplicaFailure is not easily reproducible
```
Comment (andreaTP) [inline operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:117]: Given this comment, should we iterate over this logic? (in a separate ticket)
https://github.com/keycloak/keycloak/blob/011d108fff1554e1784e0b9c70552e45ded7ec50/operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java#L144-L150

File: operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java [hunk 5]
```diff
@@ -122,6 +137,41 @@ private static void calculateNamespace() {
...
   }
 
+  protected static void deployDB() {
+    // DB
+    Log.info("Creating new PostgreSQL deployment");
+    k8sclient.load(KeycloakDeploymentE2EIT.class.getResourceAsStream("/example-postgres.yaml")).inNamespace(namespace).createOrReplace();
+
+    // Check DB has deployed and ready
+    Log.info("Checking Postgres is running");
+    Awaitility.await()
```
Comment (andreaTP) [inline operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java:147]: This is almost a note to self (no changes needed).
Waiting for the DB to be up get our tests in a "stable" situation, would it be better to instead "just deploy everything at once" and see if the operator is robust enough to catch up?

**Review by andreaTP — CHANGES_REQUESTED** [review]
Thanks for this work @vmuzikar ! Left a few comments 🙂

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java (hunk 1 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:128]: 👍

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java (hunk 2 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:59]: 👍

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    @Test
+    public void testServerConfiguration() {
+        try {
+            var kc = getDefaultKeycloakDeployment(k8sclient);
+            kc.getSpec().getServerConfiguration().put("KC_DB_PASSWORD", "Ay Caramba!");
+
+            deployKeycloak(k8sclient, kc, false);
+            waitForKcToFail(kc);
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:73]: 👍

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.keycloak.operator.utils.K8sUtils.deployKeycloak;
+import static org.keycloak.operator.utils.K8sUtils.getDefaultKeycloakDeployment;
+import static org.keycloak.operator.utils.K8sUtils.waitForKeycloakToBeReady;
+
+@QuarkusTest
+public class KeycloakDeploymentE2EIT extends ClusterOperatorTest {
+    @Test
+    public void given_ClusterAndOperatorRunning_when_KeycloakCRCreated_Then_KeycloakStructureIsDeployedAndStatusIsOK() {
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:25]: 👍

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import org.awaitility.Awaitility;
+import org.keycloak.operator.v2alpha1.crds.Keycloak;
+import org.keycloak.operator.v2alpha1.crds.KeycloakStatusCondition;
+
+/**
+ * @author Vaclav Muzikar <vmuzikar@redhat.com>
+ */
+public final class K8sUtils {
+    public static Keycloak getDefaultKeycloakDeployment(KubernetesClient client) {
+        return (Keycloak) client.load(K8sUtils.class.getResourceAsStream("/example-keycloak.yml")).get().get(0);
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:31]: The yaml file includes several resources so the suggested approach won't work. I choose this approach to be consistent on how we load YAMLs across the code.

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    public static void deployKeycloak(KubernetesClient client, Keycloak kc, boolean waitUntilReady) {
+        client.resources(Keycloak.class).createOrReplace(kc);
```
Comment (jonathanvila) [inline operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:35]: Where is the namespace here ? is this only deploying on the current client namespace ?

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    public static void deployDefaultKeycloak(KubernetesClient client) {
+        deployKeycloak(client, getDefaultKeycloakDeployment(client), true);
+    }
+
+    public static void waitForKeycloakToBeReady(KubernetesClient client, Keycloak kc) {
+        Log.infof("Waiting for Keycloak \"%s\"", kc.getMetadata().getName());
+        Awaitility.await()
+                .ignoreExceptions()
+                .untilAsserted(() -> {
+                    var currentKc = client.resources(Keycloak.class).withName(kc.getMetadata().getName()).get();
```
Comment (jonathanvila) [inline operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:51]: same as above, we are managing the namespace through the client ?

File: operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java (hunk 3 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java:152]: This is related to this PR as we already had try-catch in the original tests and we were printing it to the log. This change improves this as it at least allows the logs to be saved to a file.

Created:
* #10128
* https://github.com/java-operator-sdk/java-operator-sdk/issues/925

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+import java.util.List;
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.keycloak.operator.utils.K8sUtils.deployKeycloak;
+import static org.keycloak.operator.utils.K8sUtils.getDefaultKeycloakDeployment;
+import static org.keycloak.operator.utils.K8sUtils.waitForKeycloakToBeReady;
+
+@QuarkusTest
+public class KeycloakDeploymentE2EIT extends ClusterOperatorTest {
```
Comment (jonathanvila) [inline operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:23]: Considering that the namespace is created and cleaned at the Beginning and End of the Test Class having multiple tests , without knowing the order is not ensuring a deterministic test

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+            deployKeycloak(k8sclient, kc, false);
+
+            Awaitility.await()
+                    .during(Duration.ofSeconds(15)) // check if the Deployment is stable
+                    .untilAsserted(() -> {
+                        var c = k8sclient.apps().deployments().inNamespace(namespace).withName(deploymentName).get()
+                                .getSpec().getTemplate().getSpec().getContainers().get(0);
+                        assertThat(c.getImage()).isEqualTo("quay.io/keycloak/non-existing-keycloak");
+                        assertThat(c.getEnv().stream()
+                                .anyMatch(e -> e.getName().equals("KC_DB_PASSWORD") && e.getValue().equals("Ay Caramba!")))
```
Comment (jonathanvila) [inline operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:70]: jajajajaj Amazing Password !!!!! :)

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java (hunk 9 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:35]: Here:
https://github.com/keycloak/keycloak/blob/0fcf13406093b9a5a7fb2e252314548d203d8f19/operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java#L84
It's the default namespace for the client.

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java (hunk 10 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:51]: Here:
https://github.com/keycloak/keycloak/blob/0fcf13406093b9a5a7fb2e252314548d203d8f19/operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java#L84
It's the default namespace for the client.

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java (hunk 11 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:23]: Why it would not be deterministic? We're cleaning the KC deployment after each test. The namespace should be empty.

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:31]: I just found out, myself, that it's not the "recommended way" of doing it ... not fussy on this 👍

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+            Log.info("Checking Keycloak pod has ready replicas == 1");
+            assertThat(k8sclient.apps().deployments().inNamespace(namespace).withName(deploymentName).get().getStatus().getReadyReplicas()).isEqualTo(1);
+
+            // Delete CR
+            Log.info("Deleting Keycloak CR and watching cleanup");
+            k8sclient.resources(Keycloak.class).delete(kc);
+            Awaitility.await()
+                    .untilAsserted(() -> assertThat(k8sclient.apps().deployments().inNamespace(namespace).withName(deploymentName).get()).isNull());
+        } catch (Exception e) {
+            savePodLogs();
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:47]: We delete the pods after each tests.

File: operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java (hunk 5 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java:147]: deal 👍

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java (hunk 9 above)
Comment (jonathanvila) [reply operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:35]: :+1

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:31]: Which one is not "the recommended way"? 😉

File: operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java [hunk 14]
```diff
@@ -122,6 +137,41 @@ private static void calculateNamespace() {
...
+    Log.infof("Saving pod logs to %s", POD_LOGS_DIR);
+    for (var pod : k8sclient.pods().inNamespace(namespace).list().getItems()) {
+      try {
+        String podName = pod.getMetadata().getName();
+        Log.infof("Processing %s", podName);
+        String podLog = k8sclient.pods().inNamespace(namespace).withName(podName).getLog();
+        File file = new File(POD_LOGS_DIR + String.format("%s-%s.txt", namespace, podName)); // using namespace for now, if more tests fail, the log might get overwritten
+        file.getAbsoluteFile().getParentFile().mkdirs();
+        try (var fw = new FileWriter(file, false)) {
+          fw.write(podLog);
```
Comment (jonathanvila) [inline operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java:162]: what would be the point of this on CI ? Can we review these files on a failed pipeline ?

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java (hunk 12 above)
Comment (jonathanvila) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:70]: Should we have these hardcoded values as constants ? "KC_DB_PASSWORD"

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java (hunk 11 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:23]: Yes, we expect the operands to be deleted (and we actually test this). But even if they were not, we would overwrite them with the next KC deployment.

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:31]: Using the client methods just to deserialize is NOT recommended.
Using the `Serialization` utility is the recommended way.
Again, not really important for now, feel free to include or not this comment 👍

File: operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java (hunk 3 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java:152]: 👍

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+            var kc = getDefaultKeycloakDeployment(k8sclient);
+            var deploymentName = kc.getMetadata().getName();
+            deployKeycloak(k8sclient, kc, true);
+
+            kc.getSpec().setImage("quay.io/keycloak/non-existing-keycloak");
+            kc.getSpec().getServerConfiguration().put("KC_DB_PASSWORD", "Ay Caramba!");
+            deployKeycloak(k8sclient, kc, false);
+
+            Awaitility.await()
+                    .during(Duration.ofSeconds(15)) // check if the Deployment is stable
```
Comment (jonathanvila) [inline operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:64]: Why 15 seconds ? the benefit of using Awaitility is not having blocking delays and exit as soon as the value is right. 
What if in 5 seconds the value is the expected ? it will still wait 10 seconds more to do the first assert.

Even you are not specifying a final time right ? so it will be here for the eternity ?

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+    @Test
+    public void testDeploymentDurability() {
+        try {
+            var kc = getDefaultKeycloakDeployment(k8sclient);
+            var deploymentName = kc.getMetadata().getName();
+            deployKeycloak(k8sclient, kc, true);
+
+            Log.info("Trying to delete deployment");
+            assertThat(k8sclient.apps().deployments().withName(deploymentName).delete()).isTrue();
+            Awaitility.await()
```
Comment (jonathanvila) [inline operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:89]: No atMost ?

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+            var labels = Map.of("address", "EvergreenTerrace742");
+            var flandersEnvVar = new EnvVarBuilder().withName("NEIGHBOR").withValue("Stupid Flanders!").build();
+            var origSpecs = deployment.getSpec();
+
+            deployment.getMetadata().getLabels().putAll(labels);
+            deployment.getSpec().getTemplate().getSpec().getContainers().get(0).setEnv(List.of(flandersEnvVar));
+            k8sclient.apps().deployments().createOrReplace(deployment);
+
+            Awaitility.await()
+                    .untilAsserted(() -> {
```
Comment (jonathanvila) [inline operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:106]: No atMost ?

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java (hunk 12 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:70]: ☝️  as a reminder for the future :-)

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java (hunk 16 above)
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:89]: +1 ☝️

**Review by andreaTP — DISMISSED** [review]
The remaining comments are almost nitpicks, overall, LGTM

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:31]:
```java
return ((List<HasMetadata>) Serialization.unmarshal(K8sUtils.class.getResourceAsStream("/example-keycloak.yml"), Collections.emptyMap())).get(0)
```
should work ...

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java (hunk 16 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:89]: https://github.com/keycloak/keycloak/blob/0fcf13406093b9a5a7fb2e252314548d203d8f19/operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java#L172

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java (hunk 17 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:106]: https://github.com/keycloak/keycloak/blob/0fcf13406093b9a5a7fb2e252314548d203d8f19/operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java#L172

**Review by jonathanvila — DISMISSED** [review]

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java]: Ok, done. 👍

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+                        assertThat(d.getMetadata().getLabels().entrySet().containsAll(labels.entrySet())).isTrue(); // additional labels should not be overwritten
+                        assertThat(d.getSpec()).isEqualTo(origSpecs); // specs should be reconciled back to original values
+                    });
+        } catch (Exception e) {
+            savePodLogs();
+            throw e;
+        }
+    }
+
+    // TODO Test HasErrors condition; ReplicaFailure is not easily reproducible; blocked by https://github.com/keycloak/keycloak/issues/10099
```
[the remark this answers is not in the document]
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java]: 👍

File: operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java (hunk 15 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java:62]: This is to make sure the deployment is stable and no more modified back to the original values by the  reconciler loop. The timeout is testsuite wise 180 seconds.

File: operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java (hunk 14 above)
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/ClusterOperatorTest.java:162]: Created #10140.

**Review by andreaTP — DISMISSED** [review]
LGTM

**Review by andreaTP — APPROVED** [review]

**Review by jonathanvila — APPROVED** [review]
LGTM :+1:

**Review by pedroigor — APPROVED** [review]
